application/common/models/Mfa.php

```diff
@@ -66,12 +66,40 @@ public function fields(): array
         ];
     }
 
+    /**
+     * Whether this is both a new Mfa instance and already verified
+     *   (basically just for a new backup code Mfa option)
+     *  -- OR --
+     * Whether the Mfa option both had its verified value changed and is now verified
+     *
+     * @param array $changedAttributes
+     * @return bool
+     */
+    public function isNewlyVerified($insert, $changedAttributes)
+    {
+        if ($insert  && $this->verified) {
+            return true;
+        }
+        return (array_key_exists('verified', $changedAttributes) && $this->verified);
+    }
+
     public function afterSave($insert, $changedAttributes)
     {
         parent::afterSave($insert, $changedAttributes);
 
-        if ($insert) {
-            $this->sendAppropriateMessages($this->user, self::EVENT_TYPE_CREATE);
+        /*
+         *   Send an "Mfa Added email" for a new backup code option and for
+         * other types of mfa options that are newly verified
+         *
+         *   Don't send emails before they are verified, since the email will
+         * not include the most recently added option.
+         */
+        if ($this->isNewlyVerified($insert, $changedAttributes)) {
```

Comment:
I find myself jumping back and forth between this code (in `afterSave()`) and the contents of the `sendAppropriateMessages()` function to check the usage of this `isNewlyVerified()` function call.

It feels like we're letting the logic of when to send the various emails leak outside of their respective `$emailer->shouldSend....()` methods. What would you think of having `afterSave()` always call `$this->sendAppropriateMessages(...)`, and we could keep all of the conditions (for when to send each email) cleanly contained within those `shouldSend...()` methods?

I know that means `afterSave()` will need to pass along more parameters to `sendAppropriateMessages()`, but that seems an acceptable cost.

Reply:
Never mind. That you helped me see the full significance of the `EVENT_TYPE_VERIFY`, I think this is fine.